public/src/components/channelManagement/campaigns/TestDataDialog.tsx

```diff
@@ -0,0 +1,537 @@
+          case 'string-block':
+            res += `${rule.label}:`;
+            if (rule.optional) {
+              res += `${val != null ? `\n` + val : ' (Not set)'}`;
+            } else {
+              res += `${val != null ? `\n` + val : ' (Data missing)'}`;
+            }
+            res += `\n\n`;
+            break;
+
+          case 'string-block-array':
+            res += `${rule.label}:`;
+            if (rule.optional && val == null) {
+              res += ' (Not set)';
+            } else if (val == null) {
+              res += ' (Data missing)';
+            } else if (!val.length) {
+              res += ' (Empty)';
+            } else {
+              val.forEach(function(v: string) {
+                res += `\n${v}`;
+              });
+            }
+            res += `\n\n`;
+            break;
+
+          case 'string-array':
+            res += `${rule.label}: `;
+            if (rule.optional && val == null) {
+              res += '(Not set)';
+            } else if (val == null) {
+              res += '(Data missing)';
+            } else if (!val.length) {
+              res += '(Empty)';
+            } else {
+              res += `${val.join(', ')}`;
+            }
+            res += `\n`;
+            break;
+
+          case 'object':
+            res += `${rule.label}:`;
+            if (rule.optional && val == null) {
+              res += ` (Not set)\n`;
+            } else if (val == null) {
+              res += ` (Data missing)\n`;
+            } else {
+              res += `\n`;
+              for (const [k, v] of Object.entries(val)) {
+                res += `\t${k}: ${v}\n`;
+              }
+            }
+            break;
+
+          default:
+            res += `${rule.label}: `;
+            if (rule.optional) {
+              res += `${val != null ? val : '(Not set)'}`;
+            } else {
+              res += `${val != null ? val : '(Data missing)'}`;
+            }
+            res += `\n`;
+        }
+      }
+    });
+    return res;
+  };
+
+  const parseTestVariantData = (channel: string, variants: CommonObjectInput[]): string => {
+    let res = '';
+
+    if (
+      channel == null ||
+      ![
+        'Header',
+        'Epic',
+        'EpicHoldback',
+        'EpicLiveblog',
+        'EpicAppleNews',
+        'EpicAMP',
+        'Banner1',
+        'Banner2',
+      ].includes(channel)
+    ) {
+      res += `WARNING: Test channel not recognised!`;
+    } else if (!variants.length) {
+      res += `WARNING: no variants have been created for this Test!`;
+    } else if (['Header'].includes(channel)) {
+      variants.forEach(v => {
+        res += `Variant: ${v.name}
+---------------------------------------------------------------------
+${parseData(channel, variantFields.core, v)}
+Copy and CTAs
+~~~~~~~~~~~~~~~~~~~~~~~~
+${parseData(channel, variantFields.copy, v.content)}`;
+
+        if (v.mobileContent == null) {
+          res += `
+Mobile copy and CTAs
```

Comment:
You could move this two lines out of the `if` and not repeat them

Reply:
I did it this way because Epic variants don't have the capacity to display different messaging or CTAs on mobile devices, unlike Banners and Headers. (This is different to the device targeting issue you spotted above)